### PR TITLE
Arm64: Align methods containing loops to 32B

### PR DESCRIPTION
**Pull request by kunalspathak**
From [Arm Neoverse N1 Software Optimization Guide](https://developer.arm.com/documentation/swog309707/a) it is beneficial to align methods at 32B boundary. This will further work well when we start loop alignment at 32B boundaries.

image

As a continuation of https://github.com/dotnet/runtime/pull/2249, align bigger methods having loops to 32B alignment boundary.

- I have also added a tighter check on the code size we emit for Arm64 to make sure that there is no misestimation happening for Arm64. 
- There was an assert that would validate the `idCodeSize()` before printing the instruction which was necessary because the callers never pass the `sz` and it

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib , @jkotas

**Comment by echesakovMSFT** [comment]
@kunalspathak Have we ever seen an overestimation issue on Arm64? I thought we shouldn't given that all the instructions have the same size.

**Comment by kunalspathak** [comment]
> Have we ever seen an overestimation issue on Arm64? I thought we shouldn't

That's right, but the assert we had was little lose. I updated it for Arm64 to make sure that the memory consumed size and the code size matches.

**Review by echesakov — APPROVED** [review]
LGTM.

If I read this right, we don't attempt to align methods with R2R?

File: src/coreclr/jit/emit.cpp [hunk 1]
```diff
@@ -5657,8 +5657,8 @@ unsigned emitter::emitEndCodeGen(Compiler* comp,
     }
 #endif
 
-#ifdef TARGET_XARCH
-    // For x64/x86, align methods that are "optimizations enabled" to 32 byte boundaries if
+#if defined(TARGET_XARCH) || defined(TARGET_ARM64)
+    // For x64/x86/arm64, align methods that are "optimizations enabled" to 32 byte boundaries if
```
Comment (EgorBo) [inline src/coreclr/jit/emit.cpp:5661]: Btw, does it make any sense to align methods without loops in terms of performance? I just wonder if we should also enable it for "hot" methods with PGO/callcount data

**Comment by kunalspathak** [comment]
> If I read this right, we don't attempt to align methods with R2R?

Correct, that is a separate work item.

File: src/coreclr/jit/emit.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/emit.cpp:5661]: We could do that but we need to validate if it pays off. Remember that aligning methods to higher boundary is not free and introduces memory fragmentation.

**Review by BruceForstall — APPROVED** [review]

**Comment by kunalspathak** [comment]
Failures unrelated.

**Pull request merged**